<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/07_baseline_rf.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NB07 — Baseline Supervisionado sobre Estados Operacionais

## 1. Contexto

Após a construção da base rotulada no NB06, este notebook estabelece um baseline
supervisionado sobre os estados operacionais da série temporal.

Seu foco não é a tarefa final de antecipação binária em horizonte futuro — que será
tratada nos notebooks posteriores —, mas sim a validação estrutural de que a base
rotulada contém sinal suficiente para aprendizado supervisionado coerente.

Nesta etapa, a pergunta central é:

**os atributos construídos até aqui são suficientes para separar, de forma razoável,
regimes operacionais distintos da série?**

## 2. Objetivo

O objetivo deste notebook é:
- construir um baseline supervisionado utilizando Random Forest;
- avaliar a separabilidade entre regime normal e regime de evento;
- avaliar a separabilidade entre as quatro classes de estado;
- controlar explicitamente o uso de tempo absoluto como feature;
- registrar métricas de referência para comparação com as etapas posteriores do pipeline.

## 3. Papel no Pipeline Atualizado

Este notebook funciona como uma etapa de validação estrutural do pipeline.

Ele conecta:
- os episódios oficiais e estados derivados no NB06;
- a base de atributos produzida no NB05;
- a avaliação supervisionada preliminar sobre a base rotulada.

Diferentemente do NB08, este notebook ainda não usa a formulação final de antecipação
em horizonte futuro como tarefa principal. Em vez disso, verifica se a estrutura rotulada
da série já é compatível com aprendizado supervisionado consistente.

## 4. Estrutura Experimental

A avaliação é conduzida em dois cenários de features:

### 4.1 Cenário A — com tempo absoluto
Inclui `bucket_start_us` entre as variáveis explicativas.

### 4.2 Cenário B — sem tempo absoluto
Remove `bucket_start_us`, permitindo verificar se o modelo está aprendendo padrões
estruturais do comportamento da série ou apenas um regime temporal absoluto.

Essa comparação é importante porque a base rotulada é ordenada no tempo e pode refletir
mudanças de regime ao longo da série.

## 5. Tarefas Avaliadas

Serão avaliadas duas tarefas:

### 5.1 Classificação binária
`evento` versus `NORMAL`, em que:
- `evento = BEFORE ∪ DURING ∪ AFTER`
- `NORMAL = NORMAL`

### 5.2 Classificação multiclasse
Quatro classes:
- `NORMAL`
- `BEFORE`
- `DURING`
- `AFTER`

A tarefa binária testa separabilidade ampla entre regime normal e demais estados.
A tarefa multiclasse testa a granularidade estrutural da rotulagem produzida no NB06.

## 6. Seleção de Features

Nesta versão, a seleção de features é feita por **lista explícita de variáveis permitidas**,
e não por simples inclusão de todas as colunas numéricas.

Essa escolha evita que:
- colunas auxiliares de horizonte futuro;
- identificadores derivados da formulação supervisionada;
- campos de transição nominal;
- ou artefatos introduzidos em versões futuras do pipeline

entrem acidentalmente como variáveis de entrada do modelo.

## 7. Estratégia de Avaliação

A avaliação é realizada com:
- split temporal fixo (80% treino / 20% teste), preservando ordem cronológica;
- comparação entre cenários com e sem tempo absoluto;
- validação temporal complementar com `TimeSeriesSplit`, para observar estabilidade
  estrutural do baseline.

## 8. Métricas de Interesse

As métricas utilizadas incluem:

### 8.1 Classificação binária
- Accuracy
- Precision
- Recall
- F1-score
- ROC-AUC

### 8.2 Classificação multiclasse
- Accuracy
- Macro-F1
- Weighted-F1

Também são registradas:
- matrizes de confusão;
- relatórios de classificação;
- distribuição das classes por treino e teste;
- desempenho comparativo entre cenários com e sem tempo absoluto.

## 9. Saídas Esperadas

Ao final deste notebook, espera-se obter:
- métricas de baseline para classificação binária;
- métricas de baseline para classificação multiclasse;
- comparação direta entre cenários com e sem `bucket_start_us`;
- modelos salvos para cada cenário;
- resumo estruturado do experimento.

## 10. Relação com as Próximas Etapas

Os resultados deste notebook orientam:
- a interpretação da qualidade estrutural da base rotulada;
- a análise de eventual viés de regime temporal absoluto;
- a preparação para a tarefa supervisionada principal do NB08;
- a consolidação experimental posterior.

## 11. Observações Metodológicas

Este notebook não substitui o problema principal de antecipação formulado no NB06.

Seu papel é fornecer uma verificação experimental intermediária:
- se a base rotulada tem coerência estrutural;
- se as features capturam sinal operacional relevante;
- e se o uso de tempo absoluto altera sensivelmente o baseline.

Essa etapa é importante para separar:
- aprendizado estrutural do comportamento da série;
- de eventual aprendizado espúrio associado à posição temporal absoluta.

In [1]:
# ============================================================
# 07_baseline_rf.ipynb
# Baseline supervisionado com Random Forest
# Binário (evento vs NORMAL) + multiclasse (NORMAL/BEFORE/DURING/AFTER)
# Comparação controlada: com e sem bucket_start_us
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------
from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import json

import numpy as np
import pandas as pd
from IPython.display import display

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull).")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, MODELS_PATH, ensure_dirs
ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH =", REPORTS_PATH)
print("MODELS_PATH =", MODELS_PATH)

def log(msg: str) -> None:
    print(f"[07_baseline_rf] {msg}")

# -----------------------------
# 1) IMPORTS ML
# -----------------------------
import joblib
from sklearn.model_selection import TimeSeriesSplit
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_recall_fscore_support,
    classification_report,
    confusion_matrix,
    roc_auc_score
)

# -----------------------------
# 2) METADADOS DO NB06
# -----------------------------
NB06_SUMMARY_FILE = REPORTS_PATH / "06_labeling_states_summary.json"

nb06_summary = {}
if NB06_SUMMARY_FILE.exists():
    nb06_summary = json.loads(NB06_SUMMARY_FILE.read_text(encoding="utf-8"))
    log(f"Resumo do NB06 carregado: {NB06_SUMMARY_FILE}")
else:
    log("Resumo do NB06 não encontrado; seguindo sem metadados adicionais.")

SCENARIO_LABEL = nb06_summary.get("scenario_label", "unknown")
REMOVE_HOUR_ZERO = nb06_summary.get("remove_hour_zero", None)
EXPECTED_ROWS_LABELED = nb06_summary.get("rows_labeled", None)
EXPECTED_STATE_DIST = nb06_summary.get("states_distribution", None)

# -----------------------------
# 3) CARREGAR DATASET ROTULADO
# -----------------------------
DATA_FILE = FEATURES_PATH / "window_5min_labeled.parquet"
assert DATA_FILE.exists(), f"Arquivo não encontrado: {DATA_FILE}"

df = pd.read_parquet(DATA_FILE).sort_values("bucket_id").reset_index(drop=True)

required_cols = {"bucket_id", "state", "fail_rate"}
missing = required_cols - set(df.columns)
assert not missing, f"Colunas ausentes em labeled dataset: {missing}"

assert df["bucket_id"].is_monotonic_increasing, "bucket_id não está ordenado."
assert df["bucket_id"].is_unique, "bucket_id não é único."

if EXPECTED_ROWS_LABELED is not None:
    assert int(len(df)) == int(EXPECTED_ROWS_LABELED), (
        f"Quantidade de linhas difere do NB06: "
        f"esperado={EXPECTED_ROWS_LABELED}, observado={len(df)}"
    )

dist_total = df["state"].value_counts().to_dict()
log(f"Dataset rotulado: shape={df.shape}")
log(f"Distribuição total: {dist_total}")

if EXPECTED_STATE_DIST is not None:
    expected_norm = {str(k): int(v) for k, v in EXPECTED_STATE_DIST.items()}
    observed_norm = {str(k): int(v) for k, v in dist_total.items()}
    assert observed_norm == expected_norm, (
        f"Distribuição de estados difere do NB06: "
        f"esperado={expected_norm}, observado={observed_norm}"
    )

# -----------------------------
# 4) SCHEMA HARDENING
# -----------------------------
# Lista explícita de features permitidas.
# Tudo o que não estiver aqui fica automaticamente fora.
ALLOWED_FEATURES = [
    "bucket_start_us",
    "n_events",
    "n_failed",
    "n_machines",
    "n_collections",
    "mean_priority",
    "mean_req_cpus",
    "mean_req_mem",
    "req_cpus_presence_rate",
    "req_mem_presence_rate",
    "event_FAIL_count",
    "event_SCHEDULE_count",
    "event_FINISH_count",
    "event_ENABLE_count",
    "event_LOST_count",
    "event_EVICT_count",
    "event_KILL_count",
    "fail_rate",
    "is_empty_window",
    "lag_1",
    "lag_2",
    "lag_3",
    "rolling_mean_1h",
    "rolling_std_1h",
    "pct_change",
    "zscore_expanding",
]

missing_allowed = [c for c in ALLOWED_FEATURES if c not in df.columns]
assert not missing_allowed, f"Features esperadas ausentes no dataset: {missing_allowed}"

numeric_cols_all = df.select_dtypes(include=[np.number]).columns.tolist()
allowed_features_present = [c for c in ALLOWED_FEATURES if c in numeric_cols_all]
excluded_numeric_cols = [c for c in numeric_cols_all if c not in allowed_features_present]

log(f"Features permitidas presentes: {len(allowed_features_present)}")
log(f"Lista permitida: {allowed_features_present}")
log(f"Colunas numéricas excluídas por schema hardening: {excluded_numeric_cols}")

# Sanidade: colunas claramente auxiliares não podem entrar acidentalmente
forbidden_prefixes = ("future_", "next_", "transition_")
forbidden_exact = {
    "bucket_id",
    "is_critical",
    "prediction_horizon_minutes",
}

for col in allowed_features_present:
    assert not any(col.startswith(p) for p in forbidden_prefixes), \
        f"Coluna proibida entrou nas features: {col}"
    assert col not in forbidden_exact, \
        f"Coluna proibida entrou nas features: {col}"

# -----------------------------
# 5) PREPARAÇÃO DOS ALVOS
# -----------------------------
LABELS_MULTI = ["NORMAL", "BEFORE", "DURING", "AFTER"]

y_multi = df["state"].astype(str).copy()
y_bin = (y_multi != "NORMAL").astype(int)

# -----------------------------
# 6) FUNÇÕES AUXILIARES
# -----------------------------
TEST_RATIO = 0.20
RUN_TSCV = True
N_SPLITS_TSCV = 5

def get_feature_set(use_time_absolute: bool):
    cols = allowed_features_present.copy()
    if not use_time_absolute:
        cols = [c for c in cols if c != "bucket_start_us"]
    return cols

def fixed_time_split(X, yb, ym):
    n = len(X)
    test_size = int(np.ceil(n * TEST_RATIO))
    train_end = n - test_size

    X_train = X.iloc[:train_end].copy()
    X_test = X.iloc[train_end:].copy()

    y_train_bin = yb.iloc[:train_end].copy()
    y_test_bin = yb.iloc[train_end:].copy()

    y_train_multi = ym.iloc[:train_end].copy()
    y_test_multi = ym.iloc[train_end:].copy()

    return {
        "n_total": int(n),
        "n_train": int(len(X_train)),
        "n_test": int(len(X_test)),
        "train_end": int(train_end),
        "X_train": X_train,
        "X_test": X_test,
        "y_train_bin": y_train_bin,
        "y_test_bin": y_test_bin,
        "y_train_multi": y_train_multi,
        "y_test_multi": y_test_multi,
    }

def build_rf_binary():
    return RandomForestClassifier(
        n_estimators=300,
        random_state=SEED,
        n_jobs=-1,
        class_weight="balanced_subsample"
    )

def build_rf_multi():
    return RandomForestClassifier(
        n_estimators=400,
        random_state=SEED,
        n_jobs=-1,
        class_weight="balanced_subsample"
    )

def eval_binary_fixed(X_train, X_test, y_train_bin, y_test_bin):
    clf = build_rf_binary()
    clf.fit(X_train, y_train_bin)

    pred = clf.predict(X_test)
    proba = clf.predict_proba(X_test)[:, 1] if hasattr(clf, "predict_proba") else None

    acc = accuracy_score(y_test_bin, pred)
    prec, rec, f1, _ = precision_recall_fscore_support(
        y_test_bin, pred, average="binary", zero_division=0
    )

    auc = None
    if proba is not None and len(np.unique(y_test_bin)) == 2:
        try:
            auc = float(roc_auc_score(y_test_bin, proba))
        except Exception:
            auc = None

    cm = confusion_matrix(y_test_bin, pred).tolist()

    return clf, {
        "acc": float(acc),
        "precision": float(prec),
        "recall": float(rec),
        "f1": float(f1),
        "roc_auc": auc,
        "confusion_matrix": cm,
    }

def eval_multi_fixed(X_train, X_test, y_train_multi, y_test_multi):
    clf = build_rf_multi()
    clf.fit(X_train, y_train_multi)

    pred = clf.predict(X_test)

    acc = accuracy_score(y_test_multi, pred)
    macro_prec, macro_rec, macro_f1, _ = precision_recall_fscore_support(
        y_test_multi, pred, average="macro", zero_division=0
    )
    w_prec, w_rec, w_f1, _ = precision_recall_fscore_support(
        y_test_multi, pred, average="weighted", zero_division=0
    )

    cm = confusion_matrix(y_test_multi, pred, labels=LABELS_MULTI).tolist()
    report_text = classification_report(
        y_test_multi, pred, labels=LABELS_MULTI, zero_division=0
    )

    return clf, {
        "acc": float(acc),
        "macro_f1": float(macro_f1),
        "weighted_f1": float(w_f1),
        "macro_precision": float(macro_prec),
        "macro_recall": float(macro_rec),
        "weighted_precision": float(w_prec),
        "weighted_recall": float(w_rec),
        "confusion_matrix": cm,
        "labels_order": LABELS_MULTI,
        "classification_report_text": report_text,
    }

def eval_tscv_binary(X, yb, n_splits=5):
    tscv = TimeSeriesSplit(n_splits=n_splits)
    folds = []

    for fold, (tr, te) in enumerate(tscv.split(X), start=1):
        clf = build_rf_binary()
        clf.fit(X.iloc[tr], yb.iloc[tr])

        pred = clf.predict(X.iloc[te])
        proba = clf.predict_proba(X.iloc[te])[:, 1] if hasattr(clf, "predict_proba") else None

        acc = accuracy_score(yb.iloc[te], pred)
        prec, rec, f1, _ = precision_recall_fscore_support(
            yb.iloc[te], pred, average="binary", zero_division=0
        )

        auc = None
        if proba is not None and len(np.unique(yb.iloc[te])) == 2:
            try:
                auc = float(roc_auc_score(yb.iloc[te], proba))
            except Exception:
                auc = None

        folds.append({
            "fold": int(fold),
            "n_train": int(len(tr)),
            "n_test": int(len(te)),
            "acc": float(acc),
            "precision": float(prec),
            "recall": float(rec),
            "f1": float(f1),
            "roc_auc": auc,
        })

    return folds

def eval_tscv_multiclass(X, ym, n_splits=5):
    tscv = TimeSeriesSplit(n_splits=n_splits)
    folds = []

    for fold, (tr, te) in enumerate(tscv.split(X), start=1):
        clf = build_rf_multi()
        clf.fit(X.iloc[tr], ym.iloc[tr])

        pred = clf.predict(X.iloc[te])

        acc = accuracy_score(ym.iloc[te], pred)
        macro_prec, macro_rec, macro_f1, _ = precision_recall_fscore_support(
            ym.iloc[te], pred, average="macro", zero_division=0
        )
        w_prec, w_rec, w_f1, _ = precision_recall_fscore_support(
            ym.iloc[te], pred, average="weighted", zero_division=0
        )

        folds.append({
            "fold": int(fold),
            "n_train": int(len(tr)),
            "n_test": int(len(te)),
            "acc": float(acc),
            "macro_f1": float(macro_f1),
            "weighted_f1": float(w_f1),
            "macro_precision": float(macro_prec),
            "macro_recall": float(macro_rec),
            "weighted_precision": float(w_prec),
            "weighted_recall": float(w_rec),
        })

    return folds

def summarize_tscv(folds, metric_names):
    out = {"folds": folds}
    for metric in metric_names:
        vals = [f[metric] for f in folds if f.get(metric) is not None]
        out[f"{metric}_mean"] = float(np.mean(vals)) if vals else None
        out[f"{metric}_std"] = float(np.std(vals, ddof=0)) if vals else None
    return out

# -----------------------------
# 7) EXECUÇÃO DOS CENÁRIOS
# -----------------------------
scenario_results = {}

for scenario_name, use_time_absolute in {
    "with_time_absolute": True,
    "without_time_absolute": False,
}.items():
    log(f"=== CENÁRIO: {scenario_name} ===")

    X_cols = get_feature_set(use_time_absolute=use_time_absolute)
    assert len(X_cols) > 0, "Nenhuma feature selecionada para o cenário."

    X = df[X_cols].copy()

    split = fixed_time_split(X, y_bin, y_multi)

    dist_train_multi = split["y_train_multi"].value_counts().to_dict()
    dist_test_multi = split["y_test_multi"].value_counts().to_dict()

    dist_train_bin = split["y_train_bin"].value_counts().sort_index().to_dict()
    dist_test_bin = split["y_test_bin"].value_counts().sort_index().to_dict()

    log(f"Features ({len(X_cols)}): {X_cols}")
    log(f"Split temporal: train={split['n_train']} test={split['n_test']}")
    log(f"Distribuição treino multi: {dist_train_multi}")
    log(f"Distribuição teste  multi: {dist_test_multi}")

    # Fixed split
    rf_bin, binary_res = eval_binary_fixed(
        split["X_train"], split["X_test"],
        split["y_train_bin"], split["y_test_bin"]
    )

    rf_multi, multi_res = eval_multi_fixed(
        split["X_train"], split["X_test"],
        split["y_train_multi"], split["y_test_multi"]
    )

    log(
        f"[BIN][{scenario_name}] "
        f"acc={binary_res['acc']:.4f} "
        f"prec={binary_res['precision']:.4f} "
        f"rec={binary_res['recall']:.4f} "
        f"f1={binary_res['f1']:.4f} "
        f"auc={binary_res['roc_auc'] if binary_res['roc_auc'] is not None else 'NA'}"
    )

    log(
        f"[MULTI][{scenario_name}] "
        f"acc={multi_res['acc']:.4f} "
        f"macro_f1={multi_res['macro_f1']:.4f} "
        f"weighted_f1={multi_res['weighted_f1']:.4f}"
    )

    # Optional TSCV
    tscv_binary = None
    tscv_multiclass = None

    if RUN_TSCV:
        folds_bin = eval_tscv_binary(X, y_bin, n_splits=N_SPLITS_TSCV)
        folds_multi = eval_tscv_multiclass(X, y_multi, n_splits=N_SPLITS_TSCV)

        tscv_binary = summarize_tscv(
            folds_bin,
            metric_names=["acc", "precision", "recall", "f1", "roc_auc"]
        )
        tscv_multiclass = summarize_tscv(
            folds_multi,
            metric_names=[
                "acc", "macro_f1", "weighted_f1",
                "macro_precision", "macro_recall",
                "weighted_precision", "weighted_recall"
            ]
        )

        log(
            f"[TSCV-BIN][{scenario_name}] "
            f"f1_mean={tscv_binary['f1_mean']:.4f} "
            f"auc_mean={tscv_binary['roc_auc_mean'] if tscv_binary['roc_auc_mean'] is not None else 'NA'}"
        )
        log(
            f"[TSCV-MULTI][{scenario_name}] "
            f"macro_f1_mean={tscv_multiclass['macro_f1_mean']:.4f} "
            f"weighted_f1_mean={tscv_multiclass['weighted_f1_mean']:.4f}"
        )

    # Save models
    bin_model_path = MODELS_PATH / f"rf_binary_{scenario_name}.joblib"
    multi_model_path = MODELS_PATH / f"rf_multiclass_{scenario_name}.joblib"

    joblib.dump(rf_bin, bin_model_path)
    joblib.dump(rf_multi, multi_model_path)

    scenario_results[scenario_name] = {
        "use_time_absolute": bool(use_time_absolute),
        "n_features": int(len(X_cols)),
        "features": X_cols,
        "split": {
            "n_total": int(split["n_total"]),
            "n_train": int(split["n_train"]),
            "n_test": int(split["n_test"]),
            "class_distribution_train_multi": {k: int(v) for k, v in dist_train_multi.items()},
            "class_distribution_test_multi": {k: int(v) for k, v in dist_test_multi.items()},
            "class_distribution_train_bin": {str(k): int(v) for k, v in dist_train_bin.items()},
            "class_distribution_test_bin": {str(k): int(v) for k, v in dist_test_bin.items()},
        },
        "binary_fixed": binary_res,
        "multiclass_fixed": multi_res,
        "binary_tscv": tscv_binary,
        "multiclass_tscv": tscv_multiclass,
        "models": {
            "binary_path": str(bin_model_path),
            "multiclass_path": str(multi_model_path),
        }
    }

# -----------------------------
# 8) COMPARAÇÃO DIRETA
# -----------------------------
comparison = {
    "binary_fixed_f1_delta_with_minus_without_time": (
        scenario_results["with_time_absolute"]["binary_fixed"]["f1"] -
        scenario_results["without_time_absolute"]["binary_fixed"]["f1"]
    ),
    "binary_fixed_auc_delta_with_minus_without_time": (
        (scenario_results["with_time_absolute"]["binary_fixed"]["roc_auc"] or 0.0) -
        (scenario_results["without_time_absolute"]["binary_fixed"]["roc_auc"] or 0.0)
    ),
    "multiclass_fixed_macro_f1_delta_with_minus_without_time": (
        scenario_results["with_time_absolute"]["multiclass_fixed"]["macro_f1"] -
        scenario_results["without_time_absolute"]["multiclass_fixed"]["macro_f1"]
    ),
    "multiclass_fixed_weighted_f1_delta_with_minus_without_time": (
        scenario_results["with_time_absolute"]["multiclass_fixed"]["weighted_f1"] -
        scenario_results["without_time_absolute"]["multiclass_fixed"]["weighted_f1"]
    ),
}

log(f"Comparação direta: {comparison}")

# -----------------------------
# 9) SUMMARY JSON
# -----------------------------
summary = {
    "scenario_label": SCENARIO_LABEL,
    "remove_hour_zero": REMOVE_HOUR_ZERO,
    "input_labeled_file": str(DATA_FILE),
    "rows_total": int(len(df)),
    "cols_total": int(df.shape[1]),
    "states_distribution_total": {k: int(v) for k, v in dist_total.items()},
    "allowed_features": ALLOWED_FEATURES,
    "allowed_features_present": allowed_features_present,
    "excluded_numeric_cols_by_schema_hardening": excluded_numeric_cols,
    "run_tscv": bool(RUN_TSCV),
    "n_splits_tscv": int(N_SPLITS_TSCV),
    "test_ratio": TEST_RATIO,
    "labels_order_multiclass": LABELS_MULTI,
    "scenarios": scenario_results,
    "comparison": comparison,
}

SUMMARY_FILE = REPORTS_PATH / "07_baseline_rf_summary.json"
SUMMARY_SCENARIO_FILE = REPORTS_PATH / f"07_baseline_rf_summary_{SCENARIO_LABEL}.json"

SUMMARY_FILE.write_text(json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8")
SUMMARY_SCENARIO_FILE.write_text(json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8")

log(f"Resumo salvo (canônico): {SUMMARY_FILE}")
log(f"Resumo salvo (cenário): {SUMMARY_SCENARIO_FILE}")
log("Notebook 07 finalizado com sucesso.")

# -----------------------------
# 10) SAÍDAS DE CONFERÊNCIA
# -----------------------------
print("\n=== DISTRIBUIÇÃO TOTAL DE ESTADOS ===")
display(pd.Series(dist_total, name="count").to_frame())

for scenario_name, res in scenario_results.items():
    print(f"\n=== CENÁRIO: {scenario_name} ===")
    print(f"Features ({res['n_features']}): {res['features']}")

    print("\n[BINÁRIO - FIXED]")
    print(res["binary_fixed"])

    print("\n[MULTICLASSE - FIXED]")
    print(res["multiclass_fixed"]["classification_report_text"])
    print("Confusion matrix:", res["multiclass_fixed"]["confusion_matrix"])

    if RUN_TSCV:
        print("\n[BINÁRIO - TSCV]")
        print(res["binary_tscv"])

        print("\n[MULTICLASSE - TSCV]")
        print(res["multiclass_tscv"])

print("\n=== COMPARAÇÃO DIRETA ===")
print(comparison)

Mounted at /content/drive
[Bootstrap] Atualizando repositório (git pull).
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH = /content/drive/MyDrive/Mestrado/04-reports
MODELS_PATH = /content/drive/MyDrive/Mestrado/03-models
[07_baseline_rf] Resumo do NB06 carregado: /content/drive/MyDrive/Mestrado/04-reports/06_labeling_states_summary.json
[07_baseline_rf] Dataset rotulado: shape=(8927, 39)
[07_baseline_rf] Distribuição total: {'NORMAL': 5572, 'BEFORE': 1987, 'AFTER': 979, 'DURING': 389}
[07_baseline_rf] Features permitidas presentes: 26
[07_baseline_rf] Lista permitida: ['bucket_start_us', 'n_events', 'n_failed', 'n_machines', 'n_collections', 'mean_priority', 'mean_req_cpus', 'mean_req_mem', 'req_cpus_presence_rate', 'req_mem_presence_rate', 'event_FAIL_count', 'event_SCHEDULE_count', 'event_FINISH_count', 'event_ENABLE_count', 'event_LOST_count', 'event_EVICT_count', 'event_KILL_count', '

,count
NORMAL,5572
BEFORE,1987
AFTER,979
DURING,389



=== CENÁRIO: with_time_absolute ===
Features (26): ['bucket_start_us', 'n_events', 'n_failed', 'n_machines', 'n_collections', 'mean_priority', 'mean_req_cpus', 'mean_req_mem', 'req_cpus_presence_rate', 'req_mem_presence_rate', 'event_FAIL_count', 'event_SCHEDULE_count', 'event_FINISH_count', 'event_ENABLE_count', 'event_LOST_count', 'event_EVICT_count', 'event_KILL_count', 'fail_rate', 'is_empty_window', 'lag_1', 'lag_2', 'lag_3', 'rolling_mean_1h', 'rolling_std_1h', 'pct_change', 'zscore_expanding']

[BINÁRIO - FIXED]
{'acc': 0.8135498320268757, 'precision': 0.8516918646508279, 'recall': 0.9030534351145039, 'f1': 0.8766209707299, 'roc_auc': 0.8913216049778689, 'confusion_matrix': [[270, 206], [127, 1183]]}

[MULTICLASSE - FIXED]
              precision    recall  f1-score   support

      NORMAL       0.68      0.60      0.64       476
      BEFORE       0.59      0.82      0.69       805
      DURING       1.00      1.00      1.00       205
       AFTER       0.35      0.06      0.1

## 12. Conclusão da Etapa

A etapa de baseline supervisionado foi concluída com sucesso, permitindo avaliar a
separabilidade estrutural dos estados operacionais produzidos no NB06.

Nesta execução, o notebook utilizou:
- a base rotulada oficial `window_5min_labeled.parquet`;
- schema hardening com seleção explícita de features permitidas;
- comparação controlada entre dois cenários:
  - com `bucket_start_us`;
  - sem `bucket_start_us`.

O experimento confirmou que a base rotulada possui sinal supervisionado relevante, tanto
na tarefa binária quanto na multiclasse. Ao mesmo tempo, mostrou que o uso de tempo
absoluto não traz ganho claro e consistente de desempenho, especialmente quando se observa
a tarefa multiclasse e a validação temporal por partições progressivas.

## 13. Achados Experimentais da Execução Atual

### 13.1 Estrutura da base utilizada

A base rotulada utilizada nesta etapa apresentou:
- `rows_total = 8927`
- `cols_total = 39`

Distribuição total dos estados:
- `NORMAL = 5572`
- `BEFORE = 1987`
- `AFTER = 979`
- `DURING = 389`

Esses valores confirmam coerência com a saída oficial do NB06 e servem como referência
para toda a avaliação realizada neste notebook.

### 13.2 Endurecimento da seleção de features

Nesta versão, a seleção de features foi feita por lista explícita de variáveis permitidas.

Foram encontradas:
- `26` features permitidas no cenário com tempo absoluto;
- `25` features permitidas no cenário sem tempo absoluto.

As colunas numéricas explicitamente excluídas por schema hardening foram:
- `bucket_id`
- `is_critical`
- `next_bucket_id`
- `future_first_during_offset`

Esse controle foi importante para evitar que variáveis auxiliares ou potencialmente
problemáticas entrassem acidentalmente no treinamento do baseline.

### 13.3 Split temporal fixo

Em ambos os cenários, o split temporal fixo resultou em:
- `n_train = 7141`
- `n_test = 1786`

Distribuição multiclasse no treino:
- `NORMAL = 5096`
- `BEFORE = 1182`
- `AFTER = 679`
- `DURING = 184`

Distribuição multiclasse no teste:
- `NORMAL = 476`
- `BEFORE = 805`
- `AFTER = 300`
- `DURING = 205`

Esse desequilíbrio temporal entre treino e teste deve ser levado em conta na interpretação
dos resultados, especialmente no cenário multiclasse.

### 13.4 Resultados do cenário com tempo absoluto

#### a) Classificação binária — split fixo

No cenário com `bucket_start_us`, o baseline binário apresentou:
- `accuracy = 0.8135`
- `precision = 0.8517`
- `recall = 0.9031`
- `f1 = 0.8766`
- `roc_auc = 0.8913`

Matriz de confusão binária:
- TN = 270
- FP = 206
- FN = 127
- TP = 1183

#### b) Classificação multiclasse — split fixo

No cenário com `bucket_start_us`, o baseline multiclasse apresentou:
- `accuracy = 0.6551`
- `macro_f1 = 0.6064`
- `weighted_f1 = 0.6124`

Comportamento por classe:
- `NORMAL`: f1 = 0.64
- `BEFORE`: f1 = 0.69
- `DURING`: f1 = 1.00
- `AFTER`: f1 = 0.10

Esse resultado mostra que a classe `DURING` foi facilmente separada, enquanto `AFTER`
continuou sendo a classe mais difícil.

### 13.5 Resultados do cenário sem tempo absoluto

#### a) Classificação binária — split fixo

No cenário sem `bucket_start_us`, o baseline binário apresentou:
- `accuracy = 0.8219`
- `precision = 0.8974`
- `recall = 0.8550`
- `f1 = 0.8757`
- `roc_auc = 0.8975`

Matriz de confusão binária:
- TN = 348
- FP = 128
- FN = 190
- TP = 1120

Em relação ao cenário com tempo absoluto:
- houve leve redução de recall;
- houve aumento de precision;
- o F1 ficou praticamente empatado;
- a ROC-AUC ficou superior.

#### b) Classificação multiclasse — split fixo

No cenário sem `bucket_start_us`, o baseline multiclasse apresentou:
- `accuracy = 0.6293`
- `macro_f1 = 0.6582`
- `weighted_f1 = 0.6244`

Comportamento por classe:
- `NORMAL`: f1 = 0.68
- `BEFORE`: f1 = 0.59
- `DURING`: f1 = 1.00
- `AFTER`: f1 = 0.35

Nesse cenário, apesar da acurácia global ligeiramente menor, a qualidade média entre
classes melhorou substancialmente, especialmente devido ao ganho expressivo da classe
`AFTER`.

### 13.6 Comparação direta entre os cenários no split fixo

A comparação direta entre os cenários gerou os seguintes deltas
(`with_time_absolute - without_time_absolute`):

- `binary_fixed_f1_delta = +0.00094`
- `binary_fixed_auc_delta = -0.00622`
- `multiclass_fixed_macro_f1_delta = -0.05173`
- `multiclass_fixed_weighted_f1_delta = -0.01201`

Esses resultados indicam que:
- no problema binário, os dois cenários ficaram muito próximos em F1;
- no binário, o cenário sem tempo absoluto teve ROC-AUC melhor;
- no multiclasse, o cenário sem tempo absoluto foi claramente superior.

Assim, não há evidência forte de que `bucket_start_us` seja necessário para o baseline
desta etapa.

### 13.7 Validação temporal por TimeSeriesSplit

#### a) Binário — médias no TSCV

Com tempo absoluto:
- `f1_mean = 0.4459`
- `roc_auc_mean = 0.8210`

Sem tempo absoluto:
- `f1_mean = 0.5829`
- `roc_auc_mean = 0.8151`

No binário, o cenário sem tempo absoluto apresentou F1 médio significativamente maior,
embora com AUC média ligeiramente inferior.

#### b) Multiclasse — médias no TSCV

Com tempo absoluto:
- `macro_f1_mean = 0.5232`
- `weighted_f1_mean = 0.6415`

Sem tempo absoluto:
- `macro_f1_mean = 0.5660`
- `weighted_f1_mean = 0.6683`

No multiclasse, o cenário sem tempo absoluto superou o cenário com tempo absoluto em
ambas as métricas principais.

### 13.8 Interpretação geral do baseline

Os resultados desta execução sugerem que:
- a base rotulada tem sinal suficiente para aprendizado supervisionado consistente;
- `DURING` é uma classe estruturalmente fácil de identificar;
- `AFTER` continua sendo a classe mais desafiadora;
- o uso de tempo absoluto não traz benefício robusto e consistente;
- a remoção de `bucket_start_us` tende a produzir comportamento mais equilibrado,
  especialmente no problema multiclasse.

Isso fortalece a leitura de que o pipeline está capturando padrões estruturais da série, e
não apenas posição temporal absoluta.

## 14. Importância para o Pipeline Analítico

O NB07 cumpre o papel de validação estrutural da base rotulada.

Ele mostra que:
- os estados definidos no NB06 são de fato aprendíveis;
- as features construídas no NB05 carregam informação útil;
- o baseline supervisionado consegue separar regimes operacionais de forma não trivial;
- a dependência de tempo absoluto não é necessária para sustentar o baseline.

Essa etapa é importante porque prepara o terreno para o NB08, onde a tarefa principal já
não será a classificação de estados, mas a antecipação binária em horizonte futuro.

## 15. Artefatos Gerados

Nesta execução, foram gerados os seguintes artefatos:

- `07_baseline_rf_summary.json`
- `07_baseline_rf_summary_keep_hour0.json`
- `rf_binary_with_time_absolute.joblib`
- `rf_multiclass_with_time_absolute.joblib`
- `rf_binary_without_time_absolute.joblib`
- `rf_multiclass_without_time_absolute.joblib`

Esses arquivos passam a constituir a referência oficial do baseline supervisionado para o
cenário `keep_hour0`.

## 16. Encaminhamento do Pipeline

Com o baseline supervisionado devidamente avaliado, o pipeline segue para:
- NB08 — Modelagem da tarefa principal de antecipação;
- NB09 — Consolidação experimental e geração das figuras finais.

A interpretação desta etapa sugere que, para os notebooks seguintes, o cenário **sem tempo
absoluto** deve ser tratado como referência preferencial, salvo justificativa explícita em
sentido contrário.